HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaESProducerE.cc

```diff
@@ -0,0 +1,58 @@
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/ESGetToken.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/ESProducer.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/ModuleFactory.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/host.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/memory.h"
+#include "HeterogeneousCore/AlpakaTest/interface/AlpakaESTestRecords.h"
+#include "HeterogeneousCore/AlpakaTest/interface/ESTestData.h"
+#include "HeterogeneousCore/AlpakaTest/interface/alpaka/AlpakaESTestData.h"
+
+#include <cmath>
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+  /**
+   * This class demonstrates an ESProducer that uses the
+   * PortableCollection-based data model, and that consumes a standard
+   * host ESProduct and converts the data into PortableCollection, and
+   * implicitly transfers the data product to device
+   */
```

Comment:
@makortel can you remind me why an ESProducer that uses this approach needs to be in the `ALPAKA_ACCELERATOR_NAMESPACE` namespace ?
Aren't all the operations happening only on the host, without any reference to the platform, device, or queues ?

Reply:
> Aren't all the operations happening only on the host, without any reference to the platform, device, or queues ?

The ESProducer indeed _looks_ like that. But under the hood, the component to copy the host-side data product to the devices of the backend is added to the system in the `ALPAKA_ACCELERATOR_NAMESPACE::ESProducer` base class.

Reply:
Mhm... OK.

So if we need to make the same ESProduct available on multiple back-ends, we need to have multiple ESProducers, and run the CPU part multiple times ?

Reply:
Yes and no. The configuration using more than one non-host backends indeed needs to have an ESProducer for each non-host backend, but out of the box the behavior would be an exception about two ESProducers producing the same host-side ESProduct. The exception can be worked around with an `ESPrefer`, that tells the framework what ESProducer to use the product. So in the end the CPU part is run only once.

Improving the situation towards making this kind of ESProducer truly independent of the backends is in my long term plan (and mentioned in https://github.com/cms-sw/cmssw/issues/40140).